roles/openafs_devel/library/openafs_install.py

```diff
@@ -0,0 +1,229 @@
+
+    :arg src: directory to copy from. must already exist
+    :arg dst: directory to copy to. created if not already present
+    :returns: a list of files/symlinks created
+    """
+    outputs = []
+    if not os.path.isdir(src):
+        raise FileError("cannot copy tree '%s': not a directory" % src)
+    try:
+        names = os.listdir(src)
+    except os.error:
+        raise FileError("error listing files in '%s'" % src)
+    if not os.path.isdir(dst):
+        os.makedirs(dst)
+    for n in names:
+        src_name = os.path.join(src, n)
+        dst_name = os.path.join(dst, n)
+        if os.path.islink(src_name):
+            link_dest = os.readlink(src_name)
+            if os.path.exists(dst_name):
+                os.remove(dst_name)
+            os.symlink(link_dest, dst_name)
+            outputs.append(dst_name)
+        elif os.path.isdir(src_name):
+            outputs.extend(copy_tree(src_name, dst_name))
+        else:
+            shutil.copy2(src_name, dst_name)
+            outputs.append(dst_name)
+    return outputs
+
+def main():
+    logger = logging.getLogger(__name__)
+    results = dict(
+        changed=False,
+        msg='',
+        ansible_facts={},
+        files=[],
+        kmods=[],
+        logfiles=[],
+    )
+    module = AnsibleModule(
+        argument_spec=dict(
+            destdir=dict(type='path', required=True),
+            logdir=dict(type='path'),
+        ),
+        supports_check_mode=False
+    )
+    destdir = module.params['destdir']
+    logdir = module.params['logdir']
+
+    if logdir and os.path.isdir(logdir):
+        install_log = os.path.join(logdir, 'install.log')
+        logging.basicConfig(
+            level=logging.INFO,
+            filename=install_log,
+            format='%(asctime)s %(levelname)s %(message)s',
+        )
+        results['logfiles'].append(install_log)
+
+    logger.info('Starting install')
+    logger.debug('Parameters: %s' % pprint.pformat(module.params))
+
+    if not os.path.isdir(destdir):
+        msg = 'destdir directory not found: %s' % destdir
+        logger.error(msg)
+        module.fail_json(msg=msg)
+
+    logger.info('Copying files from %s to /' % destdir)
+    files = copy_tree(destdir, '/')
+    results['files'] = files
+    if files:
+        results['changed'] = True
+
+    for file_ in files:
+        command = os.path.basename(file_)
+        if command in COMMANDS:
+            results['ansible_facts']['afs_' + command] = file_
+
+    if platform.system() == 'Linux':
+        for file_ in files:
+            if file_.endswith('.ko'):
+                results['kmods'].append(file_)
+
+    if logdir:
+        install_list = os.path.join(logdir, 'install.list')
+        with open(install_list, 'w') as log:
+            for file_ in files:
+                log.write('%s\n' % file_)
+        results['logfiles'].append(install_list)
+
+    if platform.system() == 'Linux':
```

Comment:
ldconfig and depmod location, etc. might be platform specific.  I wonder if it makes sense to have a "platform" class that can be tailored? (e.g. redhat/debian/solaris/etc.)

Reply:
I think any "tailoring" needs to be done in the playbook, with tasks run after the openafs_install task. This module sets facts that can be used in later tasks. For example, you'll need tasks to setup the systemd unit file on platforms that support systemd. Anything else that needs to be done in the openfs_install module can be controlled with parameters if needed.

Reply:
We probably could add support for solaris here though. If i recall we need to run something similar to ldconfig.

Reply:
ldconfig and depmod are not setting facts, but performing actions.  

I think for now this is fine, but it might be something to look at in the future to see if this could be abstracted (maybe).

Reply:
yes, i mean, we can add more facts if needed. i'd rather not abstract modules, they are meant to be concrete and parameterized. We can add parameters if needed, e.g. ldconfig command to run. thanks.